(2/4) PeerManager Unit Tests and Refactoring

julianknutsen opened the pull request:
PRs:
[(1/4) Enum Dispatch NOISE State Machine & Unit Tests](https://github.com/rust-bitcoin/rust-lightning/pull/691)
[(2/4) PeerManager Unit Tests and Refactoring](https://github.com/rust-bitcoin/rust-lightning/pull/692)
[(3/4) Conduit fixups from 494 ](https://github.com/rust-bitcoin/rust-lightning/pull/693)
[(4/4) Update OutboundQueue flush behavior and fix existing bugs that it found](https://github.com/rust-bitcoin/rust-lightning/pull/714)

[Motivation](#motivation)
[Bug Fixes](#bug-fixes)
[Interface Usage](#interface-usage)
[Refactor & Unit Tests](#testing-and-refactoring)

### Next Action: <span style="color:green">Reviewers</span>
* Review changes

#### History
* 9-14-2020 *Moved from Arik's branch. Initial basic review can be found [here](https://github.com/arik-so/rust-lightning/pull/3).*
* 9-16-2020 *Antoine completed first initial review and Julian pushed requested updates*
* 10-2-2020 *Addressed Matt's trait review and rebased against upstream*
* Update-10-5-2020 *Code handoff to @arik and @ariard for next steps*

### TL;DR
- New commits start at e4b8a59 as this builds on (1/3)
- Encapsulation of the transport layer into a separate testable object enabling better unit testing of PeerManager
- PeerManager unit tests that cover almost the entire behavior aside from sync state that is hard to test in current object design
- Bug fixes from testing and code audit
- Covers the spirit of rust-bitcoin/rust-lightning#528

### Motivation
The original rust-bitcoin/rust-lightning#494 highlighted some design issues in the way the PeerManager interacted with the Conduit and PeerState objects. In order to address the design issues and perform refactoring, the module needed to have a more complete set of tests. This PR focuses on the initial work needed to unit test PeerManager (creating the Transport abstraction), the unit tests themselves, and the resulting bugfixes and refactoring that could be performed safely with the security of the tests.

### Bug Fixes<span id="testing-and-refactoring"/>
The testing and code audit uncovered a few bugs that were fixed in the following commits:
- `fix: Fix reentrancy bug in read_event()`
- `fix: Don't allow peer_disconnect before peer_connected callbacks`
- `fix: Handle Ok(false) from route_handler callbacks in event path`
- `fix: Use a separate lock for the ephemeral_key_midstate`

### Interface Usage<span id="interface-usage"/>
Interfaces are used in this PR in a few different ways.

1. Test doubles. See the [Refactor & Unit Tests](#testing-and-refactoring) section for even more depth.
Using trait-based test doubles can make tests much easier to write and read by abstracting complex objects such as the `OutboundQueue` and replacing it with a simple `Vec<Vec<u8>>` in the tests that only care about pushing data and validating the contents. Here is an example from `transport.rs` that would otherwise have to instantiate an `OutboundQueue` for all the tests leading to more complex interactions and a larger system under test than necessary.

```
/// Trait representing a container that allows enqueuing of Vec<[u8]>
pub(super) trait IOutboundQueue {
	/// Enqueue item to the queue
	fn push_back(&mut self, item: Vec<u8>);

	/// Returns true if the queue is empty
	fn is_empty(&self) -> bool;

	/// Returns the amount of available space in queue
	fn queue_space(&self) -> size;

        ...
}

/// Implement IOutboundQueue for Vec<Vec<u8>> so it can be used as a Spy in tests
impl PayloadQueuer for Vec<Vec<u8>> {
	fn push_back(&mut self, item: Vec<u8>) {
		self.push(item)
	}

	fn is_empty(&self) -> bool {
		self.is_empty()
	}

	fn queue_space(&self) -> usize {
		unimplemented!()
	}
}

fn process_input(&mut self, input: &[u8], outbound_queue: &mut impl IOutboundQueue) -> Result<bool, String>;

#[test]
fn inbound_handshake_data_goes_to_queue() {
	let mut transport = create_inbound_for_test::<PeerHandshakeTestStubBytes>();
	let mut spy = Vec::new();

	assert_eq!(transport.process_input(&[], &mut spy), Ok(false));
	assert!(!transport.is_connected());

	assert_matches!(&spy[..], [_]);
}
```

### Testing and Refactoring<span id="testing-and-refactoring"/>
In order to test the PeerManager in an isolated way, the major dependency of the Transport layer needed to be abstracted so it could be mocked. The first set of patches deal with moving the relevant code to a new testable module that can be mocked out for the PeerManager unit tests.

The sheer number of tests required to test the PeerManager is a strong indication it should be broken down into even smaller testable objects to reduce the scope of each test. This PR is a huge step forward, but there is still plenty of work to make it more maintainable.

Refactorings were done alongside the bug fixes to make the code more maintainable. Overall, the design is much cleaner, and future features can be implemented with the safety of regression tests.

**Testing Style Overview**

Rust doesn't have a great story for dependency injection or mocking like more mature languages, so the patterns are a bit heavy-handed. The general idea is that you define a trait for each interface (dependency) the object under test needs to do its job. In the production code, it will use the full-featured "real" code, but in tests, these can be swapped out for test structures that enable more granular testing. There are generally two ways to do this that are both used in this PR.

**Static Injection**
This is the mocking used in the Transport tests. It is good for small interfaces where the number of cases is very small and requires little configuration. The entire type is swapped out in tests for a test double that has pre-defined behavior.
```
/// Stub implementation of IPeerHandshake that returns an error for process_act()
pub(super) struct PeerHandshakeTestStubFail { }

impl IPeerHandshake for PeerHandshakeTestStubFail {
	fn new_outbound(_initiator_static_private_key: &SecretKey, _responder_static_public_key: &PublicKey, _initiator_ephemeral_private_key: &SecretKey) -> Self {
		PeerHandshakeTestStubFail { }
	}

	fn set_up_outbound(&mut self) -> Vec<u8> {
		vec![]
	}

	fn new_inbound(_responder_static_private_key: &SecretKey, _responder_ephemeral_private_key: &SecretKey) -> Self {
		PeerHandshakeTestStubFail { }
	}

	fn process_act(&mut self, _input: &[u8]) -> Result<(Option<Vec<u8>>, Option<(Conduit, PublicKey)>), String> {
		Err("Oh no!".to_string())
	}
}

// Test initial states start off unconnected
#[test]
fn inbound_unconnected() {
	let transport = create_inbound_for_test::<PeerHandshakeTestStubFail>();

	assert!(!transport.is_connected());
}
```

**Dependency Inversion w/ Generics**
This is the variant used in the PeerManager tests. It is better for cases where you may need to alter the test double during the test or there are so many initial configuration cases that creating static structs and swapping them in completely doesn't make sense. In the PeerManager case, this is done through dependency inversion. The PeerManager takes in a Transport object during the initialization calls (new_inbound/new_outbound) and uses it. Generics are leveraged to avoid trait objects and default type parameters are used to clean up non-test callers.

The tests configure a Transport test double in a specific state and pass it in to test the intended behavior.

```
struct PeerManagerImpl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, TransportImpl: ITransport=Transport>

fn new_outbound_connection_with_transport(&self, descriptor: Descriptor, mut transport: TransportImpl) -> Result<Vec<u8>, PeerHandleError> {
	let mut peers = self.peers.lock().unwrap();
	let initial_bytes = transport.set_up_outbound();

	if peers.peers.insert(descriptor, Peer::<TransportImpl>::new(true, transport)).is_some() {
		panic!("PeerManager driver duplicated descriptors!");
	};
	Ok(initial_bytes)
}

// Test that a new inbound connection:
// * get_peer_node_ids() does not contain the node_id
#[test]
fn new_inbound_not_in_get_peer_node_ids() {
	let test_ctx = TestCtx::<ChannelMessageHandlerTestSpy, RoutingMessageHandlerTestStub>::new();
	let descriptor = SocketDescriptorMock::new();
	let mut transport = new_unconnected_transport!();
	let peer_manager = new_peer_manager_for_test!(&test_ctx);

	new_inbound!(peer_manager, descriptor, &mut transport);

	assert!(peer_manager.get_peer_node_ids().is_empty());
}
```

ariard reviewed (commented):
Reviewed until f008fd2, including the bug fix.

Mostly call for better documentation. I don't think there is performance regression concern due the new interface.
When it's all over a gentle documentation in `src/ln/peers/mod.rs` would be great to explain relations between interfaces and sum up design goals.

julianknutsen commented on lightning/src/ln/peers/outbound_queue.rs:
```diff
@@ -0,0 +1,350 @@
+
+impl PayloadQueuer for OutboundQueue {
+	/// Unconditionally queue item. May increase queue above soft limit.
+	fn push_back(&mut self, item: Vec<u8>) {
```
None of this code has behavior changes, but I'm glad my comments and refactoring made it more clear what the current code does.

I completely agree that this is the wrong approach, but this PR isn't about changing the behavior it is about making the code modular, testable, and fixing bugs. Then it will be safe to change.

The story in the backlog (which I'm in the process of turning into Github Issues) has a solution that involves dropping optional messages. I'll leave that to a design discussion outside this patch.

julianknutsen replied:
I changed the API to `flush()` and it flushes the entire queue. It is in a separate PR found here: https://github.com/rust-bitcoin/rust-lightning/pull/714

Unfortunately, this had to go as a separate PR at the end as it encountered an existing bug (5c4cda7) fixed in 3/3. Luckily, it also found 2 more exiting bugs so the quality bar is going up. Let me know what you think.

julianknutsen commented on lightning/src/ln/peers/handler.rs:
```diff
@@ -120,6 +123,64 @@ enum InitSyncTracker{
+	/// Write previously enqueued data to the SocketDescriptor. A return of false indicates the
+	/// underlying SocketDescriptor could not fulfill the send_data() call and the blocked state
+	/// has been set. Use unblock() when the SocketDescriptor may have more room.
+	fn try_flush_one(&mut self, descriptor: &mut impl SocketDescriptor) -> bool;
```
This is another artifact of the current code doing something strange. I'll rename it to be `try_flush_one_payload` if you think that makes it more obvious. It flushes one item that was previously enqueued which is how the name came about. Super flexible.

This should obviously move to a flush_all() behavior. It makes the sync message generation better too. But, outside of this PR as this is just refactoring and documenting current behavior so the right path can be identified.

julianknutsen replied:
I put together a patch to change the API to `flush()` and it flushes the entire queue. It is in a separate PR found here: https://github.com/rust-bitcoin/rust-lightning/pull/714

Unfortunately, this had to go as a separate PR at the end as it encountered an existing bug (5c4cda7) fixed in 3/3. Luckily, it also found 2 more exiting bugs so the quality bar is going up. Let me know what you think.

julianknutsen commented on lightning/src/ln/peers/handler.rs:
```diff
@@ -432,13 +487,14 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref> PeerManager<D
+		while !peer.pending_outbound_buffer.is_blocked() {
+			let queue_space = peer.pending_outbound_buffer.queue_space();
+			if queue_space > 0 {
 				match peer.sync_status {
```
100% agree. This is a gross design choice and I did the minimal abstraction to test the OutboundQueue while not introducing a lot of behavior changes. It is very awkward to use right now and points to a code smell.

There is a story in the backlog (currently moving into GitHub issues) where I've even called out this particular item as extremely hard to unit tests in the current design.

julianknutsen commented on lightning/src/ln/peers/handler.rs:
```diff
@@ -419,7 +419,6 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref> PeerManager<D
 			($msg: expr) => {
 				{
 					log_trace!(self.logger, "Encoding and sending sync update message of type {} to {}", $msg.type_id(), log_pubkey!(peer.their_node_id.unwrap()));
-					assert!(peer.transport.is_connected());
```
Yep that is why I removed it. At this point, `Transport::enqueue_message()` panics already.

```
fn enqueue_message<M: Encode + Writeable, Q: PayloadQueuer>(&mut self, message: &M, output_buffer: &mut Q) {
	match self.conduit {
		None => panic!("Enqueueing messages only supported after transport is connected"),
		Some(ref mut conduit) => {
			let mut buffer = VecWriter(Vec::new());
			wire::write(message, &mut buffer).unwrap();
			output_buffer.push_back(conduit.encrypt(&buffer.0));
		}
	}
}

julianknutsen commented on lightning/src/ln/peers/handler.rs:
```diff
@@ -60,6 +60,9 @@ pub(super) trait ITransport {
 	fn is_connected(&self) -> bool;
 
+	/// Returns all Messages that have been received and can be parsed by the Transport
+	fn drain_messages<L: Deref>(&mut self, logger: L) -> Result<Vec<Message>, PeerHandleError> where L::Target: Logger;
```
Done: `review: Small nits, documentation, and fixups`

julianknutsen replied:
Think this got lost in the noise.

This was the design chosen in #494 and I chose to limit the number of behavior changes in the initial refactoring to ease the review (although I'm not sure how much that really saved in the end). I've done my best to separate out behavior changes to a separate patch so they can be easily dropped if necessary.

This could just as easily be a drain iterator or an iterator followed by clear() or drain(). I don't really like the pattern of iterators that also remove, so this was the cleanest analog. I don't think there is any additional memory pressure here since the bytes exist in the Decryptor anyway and they are effectively transferred to this layer.

But, it seems reasonable to change this up in the future if there is a valid reason. An iterator/drain pattern would mesh well with my thoughts in #698.

TheBlueMatt replied:
Am I reading it wrong in `transport.rs`? It looks like it processes bytes, creating messages and copying them onto the heap (in the vec), possible with reallocation (ie copies) to grow the vec. An iterator/drain iterator would just deserialize directly onto the stack and return.

julianknutsen replied:
I think it is probably worth looking at the code after 2b8d24d and recommending your changes on top of it. At that point, the code smells around iterators that can return errors and the feature to process the undecrypted message directly from the input slice has been implemented.

The original code from 494 had a bug in that it just asserted on all errors. The initial patches fixed this up to follow a more standard iterator-that-can-error pattern, but as you will see looking at it, it is definitely a code smell and hard to maintain.

Future patches culminating in 2b8d24d made this much easier to test and reason about. If you aren't happy with the performance characteristics that is probably a good place to talk about the tradeoffs so that the right design/fix can go in.

julianknutsen commented on lightning/src/ln/peers/handshake/states.rs:
```diff
@@ -30,13 +31,6 @@ pub(super) enum HandshakeState {
 	Complete(Option<(Conduit, PublicKey)>),
 }
 
-// Trait for all individual states to implement that ensure HandshakeState::next() can
```
Done:  `review: Move IHandshakeState code move to separate patch`

julianknutsen commented on lightning/src/ln/peers/transport.rs:
```diff
@@ -0,0 +1,257 @@
+	fn process_act(&mut self, input: &[u8]) -> Result<(Option<Vec<u8>>, Option<(Conduit, PublicKey)>), String>;
+}
+
+pub(super) struct Transport<PeerHandshakeImpl: IPeerHandshake=PeerHandshake> {
```
Added a comment here. I'd like to stay away from comments that talk about what items are contained inside and more about the features they provide. Let me know if my addition seems reasonable to you.

Done: `review: Small nits, documentation, and fixups`

julianknutsen commented on lightning/src/ln/peers/handler.rs:
```diff
@@ -123,126 +187,8 @@ enum InitSyncTracker{
 }
 
-/// Trait representing a container that allows enqueuing of Vec<[u8]>
-pub(super) trait PayloadQueuer {
```
Done: `review: Move PayloadQueuer trait move to a separate commit`

julianknutsen commented on lightning/src/ln/peers/handler.rs:
```diff
@@ -120,6 +123,64 @@ enum InitSyncTracker{
+}
+
+/// Trait representing a container that can try to flush data through a SocketDescriptor
+pub(super) trait SocketDescriptorFlusher {
```
Done: `review: Small nits, documentation, and fixups`

julianknutsen commented on lightning/src/ln/peers/handler.rs:
```diff
@@ -120,6 +123,64 @@ enum InitSyncTracker{
 	NodesSyncing(PublicKey),
 }
 
+/// Trait representing a container that allows enqueuing of Vec<[u8]>
```
Done: `review: Small nits, documentation, and fixups`

julianknutsen commented on lightning/src/ln/peers/handler.rs:
```diff
@@ -40,6 +40,9 @@ use bitcoin::hashes::sha256::HashEngine as Sha256Engine;
 use ln::peers::conduit::Conduit;
+use ln::peers::outbound_queue::OutboundQueue;
+
+const MSG_BUFF_SIZE: usize = 10;
```
Done: `review: Small nits, documentation, and fixups`

julianknutsen commented:
Thanks for the progress, @ariard. I've addressed all comments and pushed a version with your requested changes.

ariard reviewed (commented):
Reviewed up until 461f3be, you can partially disregard some comments as review was in-order per-commit.

Thanks for the lot of bug fixes. I don't think there is performance regression. What I want to think for next round if this design may block us for future perf improvements (e.g a pool-thread for encryption, but even this not sure as that's symmetric crypto post-handshake which should be cheap).

julianknutsen commented on lightning/src/ln/peers/handler.rs:
```diff
@@ -65,7 +65,11 @@ pub(super) trait ITransport {
+}
 
+/// Interface PeerManager uses to queue message to send. Used primarily to restrict the interface in
+/// specific contexts. e.g. Only queueing during read_event(). No flushing allowed.
```
Done: `review: Small nits, documentation, and fixups`

julianknutsen commented on lightning/src/ln/peers/handler.rs:
```diff
@@ -181,6 +181,13 @@ struct Peer<TransportImpl: ITransport> {
 impl<TransportImpl: ITransport> Peer<TransportImpl> {
+
+	/// Returns true if an INIT message has been received from this peer. Implies that this node
+	/// can send and receive encrypted messages.
```
Done: `review: Small nits, documentation, and fixups`

julianknutsen commented on lightning/src/ln/peers/handler.rs:
```diff
@@ -214,6 +221,30 @@ struct PeerHolder<Descriptor: SocketDescriptor, TransportImpl: ITransport> {
 }
 
+impl<Descriptor: SocketDescriptor, TransportImpl: ITransport> PeerHolder<Descriptor, TransportImpl> {
+	fn initialized_peer_by_node_id(&mut self, node_id: &PublicKey) -> Option<(Descriptor, &mut Peer<TransportImpl>)> {
```
Agreed. This seems obvious to me since I've been in the code, but may not be to a new reader.

Done: `review: Small nits, documentation, and fixups`

julianknutsen commented on lightning/src/ln/peers/handler.rs:
```diff
@@ -904,105 +935,76 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, TransportImpl
+						} else {
+							//TODO: Drop the pending channel? (or just let it timeout, but that sucks)
 						}
-						self.do_attempt_write_data(&mut descriptor, &mut peer.sync_status, &mut peer.transport, &mut peer.pending_outbound_buffer);
```
The `get_peer_for_forwarding()` macro actually calls `continue` if the node isn't connected so the line is never run. One of the code smells I cleaned up with `refactor: Clean up event handling path w.r.t connected state`

Definitely not a fan of macro usages like this

julianknutsen commented on lightning/src/ln/peers/handler.rs:
```diff
@@ -1214,34 +1255,41 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, TransportImpl
+							node_id_to_descriptor.remove(&their_node_id);
+							self.message_handler.chan_handler.peer_disconnected(&their_node_id, false);
+
+							return false; // retain
```
Yep, just a case where it had to look a bit worse before it looks better. Cleaned up in `refactor: Clean up the PeerHolder access/iteration/remove` which has some comments. Let me know if that looks OK to you.

julianknutsen commented on lightning/src/ln/peers/handler.rs:
```diff
@@ -650,6 +650,85 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, TransportImpl
 	}
 
+	// Returns a valid PostInitState given a Init message
+	fn post_init_state_from_init_message(&self, init_message: &msgs::Init, their_node_id: &PublicKey) -> Result<PostInitState, PeerHandleError> {
```
`initialize_peer` seems ambiguous and it doesn't really initialize the `Peer` object in initializes the `post_init_state` object.

This helper takes an `Init Message` and returns a `post_init_state` so it felt like a transformation name was the way to go. `create_post_init_state_from_init_message()`?

julianknutsen commented on lightning/src/ln/peers/transport.rs:
```diff
@@ -82,14 +82,16 @@ impl<PeerHandshakeImpl: IPeerHandshake> ITransport for Transport<PeerHandshakeIm
 					self.their_node_id = Some(remote_pubkey);
+					Ok(true) // newly connected
+				} else {
+					Ok(false) // newly connected
```
Maybe `newly` is the wrong word? This function returns `true` on the first call that ends with a connected state. I've updated the trait API and put an inline comment in `review: Small nits, documentation, and fixups` to help future readers.

julianknutsen commented on lightning/src/ln/peers/handler.rs:
```diff
@@ -650,6 +650,85 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, TransportImpl
+	}
+
+	// Process an incoming Init message and set Peer and PeerManager state accordingly
+	fn process_init_message(&self, message: Message, descriptor: &Descriptor, peers_needing_send: &mut HashSet<Descriptor>, node_id_to_descriptor: &mut HashMap<PublicKey, Descriptor>, pending_outbound_buffer: &mut impl PayloadQueuer, outbound: bool, transport: &mut impl ITransport, post_init_state: &mut Option<PostInitState>) -> Result<(), PeerHandleError> {
```
Done: `review: Small nits, documentation, and fixups`

julianknutsen commented on lightning/src/ln/peers/handler.rs:
```diff
@@ -650,6 +650,85 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, TransportImpl
+			Message::Init(ref init_message) => {
+				log_trace!(self.logger, "Received Init message from {}", log_pubkey!(&their_node_id));
+				if node_id_to_descriptor.contains_key(&their_node_id) {
+					log_trace!(self.logger, "Got second connection with {}, closing", log_pubkey!(&their_node_id));
```
Ya, that seems reasonable. I added a review patch `review: Catch duplicate connection earlier` to catch it at the completion of the NOISE handshake instead of the INIT message processing.

julianknutsen commented on lightning/src/ln/peers/handler.rs:
```diff
@@ -650,6 +650,85 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, TransportImpl
+			log_info!(self.logger, "Peer global features required unknown version bits");
+			return Err(PeerHandleError { no_connection_possible: true }.into());
+		}
+		if init_message.features.requires_unknown_bits() {
```
review: Remove legacy (local/global) feature flag notions

julianknutsen commented on lightning/src/ln/peers/handler.rs:
```diff
@@ -650,6 +650,85 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, TransportImpl
+		}
+
+		log_info!(
+			self.logger, "Received peer Init message: data_loss_protect: {}, initial_routing_sync: {}, upfront_shutdown_script: {}, static_remote_key: {}, unknown flags (local and global): {}",
```
Done: `review: Remove legacy (local/global) feature flag notions`

julianknutsen commented:
Thanks for the full first pass @ariard. I've addressed all comments and pushed another version with your requests.

I don't think the encryption thread pool will be hard to integrate, but I'm interested to hear your thoughts. The Encryptor/Decryptor will just need to be updated be passed the pool and use it for the operations. Should make it simple to test as well since those are now separate objects.

Definitely something to do once it is a problem and not overengineer too much upfront.

julianknutsen commented:
The module had more bugs than I expected, but that validated all the work up front to make it testable. I think it is a great case study for how tests like this can be leveraged to make the code a lot more stable. You should see quick feature iteration on this layer too now that the tests exist.

julianknutsen commented:
Rebased against upstream and added licenses to new files.

TheBlueMatt commented on lightning/src/ln/peers/handler.rs:
```diff
@@ -0,0 +1,3280 @@
+use bitcoin::hashes::{HashEngine, Hash};
+use ln::peers::outbound_queue::OutboundQueue;
+use ln::peers::transport::{PayloadQueuer, Transport};
+use bitcoin::hashes::core::iter::Filter;
```
This should be `std::iter::Filter`.

julianknutsen replied:
Done: squashed into `refactor: Clean up the PeerHolder access/iteration/remove`

TheBlueMatt commented on lightning/src/ln/peers/handler.rs:
```diff
@@ -0,0 +1,3280 @@
+const OUTBOUND_QUEUE_SIZE: usize = 10;
+
+/// Interface PeerManager uses to interact with the Transport object
+pub(super) trait ITransport: MessageQueuer {
```
I find it very hard with this design to see trivially where the queues are hiding, and how big they can grow, and what each trait means. Can we get some documentation on what the difference between a Message and PayloadQueuer is, how that fits into the OutboundQueue, why `process_input` takes a trait which `OutboundQueue` implements, etc. I also think we can squash a bunch of stuff to make this much more readable - can we avoid having both a PayloadQueuer and a MessageQueuer and only have one queue and one thing that encrypts/decrypts against the queue, or is that hiding here somewhere and I dont see it?

julianknutsen replied:
Vec vs. Iterator

This was the design chosen in #494 and I chose to limit the number of behavior changes in the initial refactoring to ease the review (although I'm not sure how much that really saved in the end). I've done my best to separate out behavior changes to a separate patch so they can be easily dropped if necessary.

This could just as easily be a drain iterator or an iterator followed by `clear()` or `drain()`. I don't really like the pattern of iterators that also remove, so this was the cleanest analog. I don't think there is any additional memory pressure here since the bytes exist in the Decryptor anyway and they are effectively transferred to this layer.

But, it seems reasonable to change this up in the future if there is a valid reason. An iterator approach would mesh well with my thoughts in #698.

julianknutsen replied:
I think cleaning up the naming might make this a bit easier. I'm not attached to any of the trait names but tried to follow Rust trait idioms, but I can see how they can all blend if you haven't been staring at the code for 6 weeks straight. I've answered your questions inline. If you have better names that make the traits a lot more obvious I am happy to clean them up.

> I find it very hard with this design to see trivially where the queues are hiding,

The only queue is the `OutboundQueue` and is owned by the `PeerManager` at the top level. Same as the LL before.

>  and how big they can grow, 

The `PeerManager` shouldn't care how big they can grow and this was one of the benefits of splitting out the `OutboundQueue` object. I've kept the const at the top level, but that was only to keep a single constructor for tests and the actual implementation.

If there are performance issues in the future, the features can just be added to the `OutboundQueue` and tested in isolation.

As far as max size, the `read_event()` path pushes unconditionally. There was no behavior change for this in my patch stack and the API and docs now reflect the behavior. Is this the best long-term, probably not. That feature is captured in #703.

>and what each trait means. Can we get some documentation on what the difference between a Message and PayloadQueuer is,

A naming fixup might help the confusion. I will do my best to describe the relationships here and your feedback can help me clean up the trait documentation to help the next reader.

The `PeerHandler` owns the `OutboundQueue` which is a simple wrapper object for raw bytes that need to be sent out of the socket descriptor. The `OutboundQueue` implements two traits: 

* `PayloadQueuer` handles the raw byte enqueue into the queue
* `SocketDescriptorFlusher` handles the flushing and flow control

The `PayloadQueuer` is the interface consumed by the `Transport` layer. It needs a way to push unencrypted bytes pre-NOISE and encrypted bytes post-NOISE.

> how that fits into the OutboundQueue, why `process_input` takes a trait which `OutboundQueue` implements, etc. 

`process_input()` takes a `PayloadQueuer` because it is just responsible for writing raw bytes into the `OutboundQueue`. By taking the trait, instead of the `OutboundQueue`, it is more clear (and guaranteed by the type system) that the `Transport` layer is only responsible for pushing items and not flushing them.

> I also think we can squash a bunch of stuff to make this much more readable 

Sort of opposite feedback from @ariard who asked me to split it out more. Is this something that can just be fixed with `git diff`? I did my best to not only break up the commits into smaller consumable pieces but also walk the reviewers through my train of thought as I did the refactor. Big refactors like this don't always have an obvious endpoint when you start. You have to untangle it piece by piece and see what it looks like. When I'm reviewing large changes like this I find it adds a lot more confidence if I can see how someone approached a redesign vs. just the final product.

In the end, if you can't review it this will just sit here forever, so please let me know how I can help make it easier on you.

> - can we avoid having both a PayloadQueuer and a MessageQueuer and only have one queue and one thing that encrypts/decrypts against the queue, or is that hiding here somewhere and I don't see it?

There is one queue. The raw byte (Payload) queue `OutboundQueue` is owned by the `PeerManager`.

The `PeerManager` only interacts with `MessageQueuer` since it only needs to send post-NOISE Messages

The `Transport` only interacts with `PayloadQueuer` since it just deals with raw bytes. The unencrypted ACT bytes and encrypted Messages.

During pre-NOISE, the `Transport` layer takes in a `PayloadQueuer` during `process_input()` where it will enqueue the unencrypted ACT bytes.

During post-NOISE, the `PeerManager` asks the `Transport` layer to encrypt bytes via `enqueue_message()` which is essentially: take this Message, encrypt it, and add it to the queue.

This separation allows the `PeerManager` to own the queue and give it to the `Transport` functions that need to enqueue onto it.

So, in the end, it matches with what you want... I think. There is one queue `OutboundQueue` and `Transport` encrypts/decrypts against it.

TheBlueMatt replied:
Right, I think its possible with more documentation and trait renaming to make it clearer what's going on, maybe without ripping out *all* the test traits, though the ratio of traits-to-code seems difficult to swallow. eg it may make sense to reduce the trait count and take the complexity hit by adding more shims in the test side of the code.

julianknutsen replied:
Done: `review: Collapse OutboundQueue/Transport traits`

TheBlueMatt reviewed (commented):
I find it pretty hard to track what's going on with all the new traits, can we try to squash them a bit, and/or make the naming more consistant across traits vs structs, and/or add more documentation describing what they do?

julianknutsen commented on lightning/src/ln/peers/handler.rs:
```diff
@@ -0,0 +1,3280 @@
+
+/// Trait representing a container that can try to flush data through a SocketDescriptor. Used by the
+/// PeerManager to handle flushing the outbound queue and flow control.
+pub(super) trait SocketDescriptorFlusher {
```
Separating what an object can do and what another object needs it to provide allows for a more modular design and easier testing. The fact that `OutboundQueue` implements both interfaces is more of a convenience than a reason to flattern the interfaces.

By splitting the entire API into the interfaces that each object needs, you can easily write tests by just implementing test doubles for the subset of the API and not have to mock/spy a bunch of unused features/function. It also allows the function signatures to document the operations that the function will use which can come in handy when trying to restrict operations in different phase (such as no flush in `read_event()`)

### Testing
The transport tests for `process_input()` only need access to the subset of the API that enqueues messages for send. This allows the tests to use a simple Vec<u8> spy instead of a full OutboundQueue and get the same test coverage.

### Interface vs Object Development
Using interfaces instead of objects also focuses the attention on which features a consumer object needs to get their work done. In my experience, this leads to smaller interfaces between objects that don't overengineer anything.

Another approach I experimented with was to have a separate `SocketDescriptorFlusherImpl` object that uses composition w/ a `PayloadQueuer` and just handles the logic to add Sync messages and flow control. This separate object would implement the `SocketDescriptorFlusher` trait and fits more inline with the single responsibility principle.

Given the longer feedback loops for open source work and my experience with OO principles being less engrained in open source projects, I chose to limit the number of new objects and first see if this was an approach that was interesting to the project.

julianknutsen commented on lightning/src/ln/peers/handler.rs:
```diff
@@ -0,0 +1,3280 @@
+	CM::Target: ChannelMessageHandler,
+	RM::Target: RoutingMessageHandler,
+	L::Target: Logger {
+		inner: PeerManagerImpl<Descriptor, CM, RM, L>,
```
Using default type parameters in the public interface requires that the `Transport` type has public visibility as well. 

This would mean leaking private types as well as making it unclear from a user perspective on whether or not they needed to implement a Transport object (it would show up in the rustdoc as well). Since this is just a pattern for testability, I chose to hide it as an inner module which allowed the use of dependency injection without an interface change to the user.

julianknutsen commented:
@TheBlueMatt Thanks for taking a quick look at this. Some of the points you bring up are a little easier to talk about with concrete code. So, I am going to write a patch that uses less traits so that we can talk about the tradeoffs and see which is the right way to go moving forward.

At the time of development, I was hopeful that I would be able to continue working on LDK and really show how this type of development style can be leveraged for better testing and new features. Now that that isn't the case, I think it is more important to get it into a state that is easier for the core team to maintain while still getting a lot of the benefits from test doubles.

I'll push a new version in the next day or so when I find some time and we can go from there.

julianknutsen commented:
Rebased against upstream and addressed @TheBlueMatt's feedback in `review: Collapse OutboundQueue/Transport traits`.

I think that review patch is a good compromise between using traits for test doubles in the `OutboundQueue` and `Transport` objects, but collapsing the other interfaces to reduce complexity for those unfamiliar with trait/interfaces patterns.

TheBlueMatt commented on lightning/src/ln/peers/transport.rs:
```diff
@@ -0,0 +1,404 @@
+use std::ops::Deref;
+
+/// Interface used by Transport to interact with a handshake object
+pub trait IPeerHandshake {
```
This probably needs a rename ("I"-prefix seems to me to imply internal?) if its going to be public, as well as likely a much more thorough set of docs to explain what it does.

julianknutsen replied:
Sure. I didn't find an existing pattern for interface traits in the codebase. In previous jobs, interfaces had "I" prefixes/suffixes so the types were identifiable. Let me know what you would prefer and I can do the rename.

TheBlueMatt replied:
Hmm, right, I think there's less chance of confusion in Rust since things either need to be explicitly generic across a trait or (at least once we upgrade to newer rust) carry the `dyn` identifier. I'm not sure a prefix is required, I obviously misread the intent. In any case, does this need to be pub, or should it be pub only in fuzztarget mode/for testing? I'm not sure what the specific use-case is for the API, maybe @arik-so had one?

TheBlueMatt reviewed (commented):
Thanks, this is a bit clearer than it was, though I think there's still a lot of opportunity to reduce scoping of various structs and decouple objects/traits which are very tighly coupled right now.

julianknutsen commented:
> Thanks, this is a bit clearer than it was, though I think there's still a lot of opportunity to reduce scoping of various structs and decouple objects/traits which are very tighly coupled right now.

This feedback isn't very actionable. There is always more cleanup that can get done. The question is whether or not this is a large enough improvement from the current main or if additional PRs need to be added to get it into a more final state that you are happy merging.

Personally, I think this is a huge improvement over the preexisting code and the additional tests and bugfixes are worth merging. But, I will leave that up to the Square team to decide.

TheBlueMatt commented:
> This feedback isn't very actionable.

Right, I generally use the top-level comment to summarize my thinking (in this case across several PRs, not just this one), and inline comments to note specific action-items.

TheBlueMatt commented:
The author decided to move on some time ago, and redoing the great work here would likely require a substantial effort to rebase and address the regressions that were being worked on in the later PRs in this set when the author moved on. Going to close.
